File: computing/interview/design-patterns/behavioral/command/csharp/CommandDemo.cs
using System;
using System.Collections.Generic;

namespace Interview.DesignPatterns.Behavioral.Command
{
    public sealed class Editor
    {
        public string Content { get; private set; } = string.Empty;

        public void Append(string text)
        {
            Content += text;
        }

        public void RemoveLast(int length)
        {
            if (length <= Content.Length)
            {
                Content = Content[..^length];
            }
        }
    }

    public interface ICommand
    {
        void Execute();
        void Undo();
    }

    public sealed class AppendCommand : ICommand
    {
        private readonly Editor _editor;
        private readonly string _text;

        public AppendCommand(Editor editor, string text)
        {
            _editor = editor;
            _text = text;
        }

        public void Execute()
        {
            _editor.Append(_text);
        }

        public void Undo()
        {
            _editor.RemoveLast(_text.Length);
        }
    }

    public sealed class Invoker
    {
        private readonly Stack<ICommand> _history = new();

        public void RunCommand(ICommand command)
        {
            command.Execute();
            _history.Push(command);
        }

        public void UndoLast()
        {
            if (_history.Count == 0)
            {
                Console.WriteLine("Nothing to undo");
                return;
            }

            var command = _history.Pop();
            command.Undo();
        }
    }

    public static class Demo
    {
        public static void Main()
        {
            var editor = new Editor();
            var invoker = new Invoker();

            invoker.RunCommand(new AppendCommand(editor, "Hello"));
            invoker.RunCommand(new AppendCommand(editor, " World"));
            Console.WriteLine($"Content: {editor.Content}");

            invoker.UndoLast();
            Console.WriteLine($"After undo: {editor.Content}");
        }
    }
}

File: computing/interview/design-patterns/behavioral/command/python/command_demo.py
from __future__ import annotations
from abc import ABC, abstractmethod
from collections import deque


class Editor:
    def __init__(self) -> None:
        self.content = ""

    def append(self, text: str) -> None:
        self.content += text

    def remove_last(self, length: int) -> None:
        self.content = self.content[:-length] if length <= len(self.content) else ""


class Command(ABC):
    @abstractmethod
    def execute(self) -> None:
        ...

    @abstractmethod
    def undo(self) -> None:
        ...


class AppendCommand(Command):
    def __init__(self, editor: Editor, text: str) -> None:
        self._editor = editor
        self._text = text

    def execute(self) -> None:
        self._editor.append(self._text)

    def undo(self) -> None:
        self._editor.remove_last(len(self._text))


class Invoker:
    def __init__(self) -> None:
        self._history: deque[Command] = deque()

    def run_command(self, command: Command) -> None:
        command.execute()
        self._history.append(command)

    def undo_last(self) -> None:
        if not self._history:
            print("Nothing to undo")
            return
        command = self._history.pop()
        command.undo()


def main() -> None:
    editor = Editor()
    invoker = Invoker()

    invoker.run_command(AppendCommand(editor, "Hello"))
    invoker.run_command(AppendCommand(editor, " World"))
    print(f"Content: {editor.content}")

    invoker.undo_last()
    print(f"After undo: {editor.content}")


if __name__ == "__main__":
    main()
